# <center> US pollution analysis

### Introduction

The US Pollution Data project leverages a comprehensive dataset covering air pollution measurements across the United States — consisting of over 1.4 million observations and around 28 variables that record concentrations of major pollutants such as nitrogen dioxide (NO₂), sulfur dioxide (SO₂), carbon monoxide (CO), and ozone (O₃). 


This dataset spans several years and states, allowing detailed spatio-temporal analysis of pollutant levels, seasonal trends, and geographic variation. The primary objective of this project is to transform this raw data into clean, analysis-ready formats (e.g., Parquet), conduct exploratory data analysis (EDA) to uncover patterns and insights, and — where possible — apply machine learning techniques to forecast pollutant concentrations, classify air‑quality levels, or identify key factors driving pollution.

Given the public health importance of air quality, this project has the potential not only to improve our understanding of pollution trends in the U.S., but also to inform policy, raise awareness, or support predictive systems that warn populations about deteriorating air conditions.

The goal of this project is to:

* Clean and transform the raw data (e.g., into Parquet)
* Perform EDA to uncover trends, seasonal patterns, and geographic variation
* Apply ML techniques to forecast pollutant levels, classify air quality, and analyze feature importance

By improving our understanding of air quality trends, this project supports public health insights and data-driven policy decisions.

Dataset content:
* State Code: Numeric code representing the U.S. state
* County Code:	Numeric code for the county within the state
* Site Num:	Identifier for the air monitoring site
* Address:	Street address of the monitoring station
* State:	Full name of the U.S. state
* County:	Name of the county
* City:	City where the measurement site is located
* Date Local:	Date of the observation (YYYY-MM-DD)
* NO2 Units:	Units used for nitrogen dioxide measurements
* NO2 Mean:	Daily average NO₂ concentration
* NO2 1st Max Value:	Highest NO₂ value recorded that day
* NO2 1st Max Hour:	Hour when the highest NO₂ was recorded
* NO2 AQI:	Air Quality Index for NO₂ on that day
* O3 Units:	Units used for ozone measurements
* O3 Mean:	Daily average ozone concentration
* O3 1st Max Value:	Highest O₃ value recorded that day
* O3 1st Max Hour:	Hour when the highest O₃ was recorded
* O3 AQI:	Air Quality Index for O₃ on that day
* SO2 Units:	Units used for sulfur dioxide measurements
* SO2 Mean:	Daily average SO₂ concentration
* SO2 1st Max Value:	Highest SO₂ value recorded that day
* SO2 1st Max Hour:	Hour when the highest SO₂ was recorded
* SO2 AQI:	Air Quality Index for SO₂ on that day
* CO Units:	Units used for carbon monoxide measurements
* CO Mean:	Daily average CO concentration
* CO 1st Max Value:	Highest CO value recorded that day
* CO 1st Max Hour:	Hour when the highest CO was recorded
* CO AQI:	Air Quality Index for CO on that day

---

### 1. EDA and Initial data visualisation

First, all necessary libraries are imported

In [34]:
import os                          #import os for operating system interactions
import pandas as pd                 #import Pandas for data manipulation
import numpy as np                  #import Numpy for numerical operations
import matplotlib.pyplot as plt     #import Matplotlib for data visualization
import seaborn as sns               #import Seaborn for statistical data visualization
from plotly.subplots import make_subplots  #import Plotly subplots for creating complex figures
import plotly.express as px         #import Plotly Express for interactive visualizations
import plotly.graph_objects as go   #import Plotly Graph Objects for detailed figure customization

In [35]:
sns.set(style="whitegrid")                  # Set Seaborn style for plots
plt.rcParams["figure.figsize"] = (10,6)     # Set default figure size for Matplotlib plots

#### 1.1. ETL and EDA


In this section EDA, including data load and cleaning, is performed. As a first step, data set is loaded into DataFrame

In [36]:
df = pd.read_parquet("../data/raw/pollution_dataset.parquet", engine="pyarrow")
df.head()

Unnamed: 0  State Code  County Code  Site Num  \
0           0           4           13      3002   
1           1           4           13      3002   
2           2           4           13      3002   
3           3           4           13      3002   
4           4           4           13      3002   

                                   Address    State    County     City  \
0  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
1  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
2  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
3  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
4  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   

   Date Local          NO2 Units  ...          SO2 Units  SO2 Mean  \
0  2000-01-01  Parts per billion  ...  Parts per billion  3.000000   
1  2000-01-01  Parts per billion  ...  Parts per billion  3.000000   
2  2000-01-01  Parts per billion  ...  Parts per billion  2.975000   
3  2000-01-01  Parts per billion  ...  Parts per billion  2.975000   
4  2000-01-02  Parts per billion  ...  Parts per billion  1.958333   

   SO2 1st Max Value  SO2 1st Max Hour SO2 AQI           CO Units   CO Mean  \
0                9.0                21    13.0  Parts per million  1.145833   
1                9.0                21    13.0  Parts per million  0.878947   
2                6.6                23     NaN  Parts per million  1.145833   
3                6.6                23     NaN  Parts per million  0.878947   
4                3.0                22     4.0  Parts per million  0.850000   

   CO 1st Max Value  CO 1st Max Hour CO AQI  
0               4.2               21    NaN  
1               2.2               23   25.0  
2               4.2               21    NaN  
3               2.2               23   25.0  
4               1.6               23    NaN  

[5 rows x 29 columns]

In [37]:
df. drop(columns=['Unnamed: 0'], inplace=True)  # Drop unnecessary column
df.head()

State Code  County Code  Site Num                                  Address  \
0           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN   
1           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN   
2           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN   
3           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN   
4           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN   

     State    County     City  Date Local          NO2 Units   NO2 Mean  ...  \
0  Arizona  Maricopa  Phoenix  2000-01-01  Parts per billion  19.041667  ...   
1  Arizona  Maricopa  Phoenix  2000-01-01  Parts per billion  19.041667  ...   
2  Arizona  Maricopa  Phoenix  2000-01-01  Parts per billion  19.041667  ...   
3  Arizona  Maricopa  Phoenix  2000-01-01  Parts per billion  19.041667  ...   
4  Arizona  Maricopa  Phoenix  2000-01-02  Parts per billion  22.958333  ...   

           SO2 Units  SO2 Mean  SO2 1st Max Value SO2 1st Max Hour  SO2 AQI  \
0  Parts per billion  3.000000                9.0               21     13.0   
1  Parts per billion  3.000000                9.0               21     13.0   
2  Parts per billion  2.975000                6.6               23      NaN   
3  Parts per billion  2.975000                6.6               23      NaN   
4  Parts per billion  1.958333                3.0               22      4.0   

            CO Units   CO Mean  CO 1st Max Value CO 1st Max Hour  CO AQI  
0  Parts per million  1.145833               4.2              21     NaN  
1  Parts per million  0.878947               2.2              23    25.0  
2  Parts per million  1.145833               4.2              21     NaN  
3  Parts per million  0.878947               2.2              23    25.0  
4  Parts per million  0.850000               1.6              23     NaN  

[5 rows x 28 columns]

In the following subsection initial data set inspection is performed. Here the shape and Info of DataFrame are shown

In [38]:
print(df.shape)                     # Print the shape of the DataFrame           
print(df.info())                    # Print concise summary of the DataFrame            
print(df.dtypes)                    # Print data types of each column

(1746661, 28)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1746661 entries, 0 to 1746660
Data columns (total 28 columns):
 #   Column             Dtype  
---  ------             -----  
 0   State Code         int64  
 1   County Code        int64  
 2   Site Num           int64  
 3   Address            object 
 4   State              object 
 5   County             object 
 6   City               object 
 7   Date Local         object 
 8   NO2 Units          object 
 9   NO2 Mean           float64
 10  NO2 1st Max Value  float64
 11  NO2 1st Max Hour   int64  
 12  NO2 AQI            int64  
 13  O3 Units           object 
 14  O3 Mean            float64
 15  O3 1st Max Value   float64
 16  O3 1st Max Hour    int64  
 17  O3 AQI             int64  
 18  SO2 Units          object 
 19  SO2 Mean           float64
 20  SO2 1st Max Value  float64
 21  SO2 1st Max Hour   int64  
 22  SO2 AQI            float64
 23  CO Units           object 
 24  CO Mean            float64
 25  CO 1

In [39]:
df.dtypes.value_counts()         # Count occurrences of each data type

float64    10
int64       9
object      9
Name: count, dtype: int64

As it shown above dataset consists of 1746661 entries and 28 columns. Also dataset contains 10 float columns, 9 integer and 9 categorical columns. 

In the next steps DataFrame is checked for any incosistencies(dublicates, missing value and etc.)

In [40]:
df.isnull().sum()           # Check for missing values in each column

State Code                0
County Code               0
Site Num                  0
Address                   0
State                     0
County                    0
City                      0
Date Local                0
NO2 Units                 0
NO2 Mean                  0
NO2 1st Max Value         0
NO2 1st Max Hour          0
NO2 AQI                   0
O3 Units                  0
O3 Mean                   0
O3 1st Max Value          0
O3 1st Max Hour           0
O3 AQI                    0
SO2 Units                 0
SO2 Mean                  0
SO2 1st Max Value         0
SO2 1st Max Hour          0
SO2 AQI              872907
CO Units                  0
CO Mean                   0
CO 1st Max Value          0
CO 1st Max Hour           0
CO AQI               873323
dtype: int64

The dataset misses 872907 and 873323 values in SO2 AQI and CO AQI columns respectively. This is a big amount of missing data to just remove lines. Instead, these missing values can be calculated, as far as other columns have no missing points.

Air Quality Index (AQI) is calculated by converting measured pollutant concentrations (e.g., SO₂, CO, NO₂, O₃, PM₂.₅, PM₁₀) into a standardized scale (usually 0–500) using breakpoints.
Core AQI Formula

Each pollutant gets its own AQI number. The final AQI for the city/location is the highest of all pollutants.

Each government sets concentration ranges for each pollutant.
Example (US EPA standard):
SO₂ 1-hour breakpoints (ppb)
AQI Range	                SO₂ (ppb)
0–50 (Good)	                0–35
51–100 (Moderate)	        36–75
101–150 (Unhealthy SG)	    76–185
151–200 (Unhealthy)	        186–304
201–300 (Very Unhealthy)    305–604

CO 8-hour breakpoints (ppm)
AQI Range	CO (ppm)
0–50	    0.0–4.4
51–100	    4.5–9.4
101–150	    9.5–12.4
151–200	    12.5–15.4
201–300	    15.5–30.4

Before calculating AQIs SO2 and CO, let's confirm that dedicated cilumns do not contain negeative values (which is physically impossible)

In [41]:
neg_so2 = df[df["SO2 1st Max Value"] < 0][["SO2 Units", "SO2 Mean", "SO2 1st Max Value"]]
neg_co = df[df["CO Mean"] < 0][["CO Units", "CO Mean", "CO 1st Max Value"]]

print("Negative SO₂ values:")
print(neg_so2.head())

print("\nNegative CO values:")
print(neg_co.head())


Negative SO₂ values:
                 SO2 Units  SO2 Mean  SO2 1st Max Value
1004801  Parts per billion -0.428571               -0.3
1004802  Parts per billion -0.428571               -0.3
1091334  Parts per billion -0.391667               -0.3
1091335  Parts per billion -0.391667               -0.3
1091336  Parts per billion -0.375000               -0.3

Negative CO values:
                  CO Units   CO Mean  CO 1st Max Value
1116634  Parts per million -0.015385               0.0
1116636  Parts per million -0.015385               0.0
1116638  Parts per million -0.019048               0.1
1116639  Parts per million -0.009524               0.1
1116640  Parts per million -0.019048               0.1


In [42]:
print(neg_co.value_counts().sum())
print(neg_so2.value_counts().sum())

1064
8286


There are 1064 and 8286 negative values in CO Mean and SO2 1st Max Value columns, respectively. In this case, removing these values from the dataset is the simplest, cleanest and safest approach.

Negative pollution values are invalid. SO₂ and CO cannot be negative in reality. These values come from:
* sensor malfunction
* data ingestion error
* interpolation issues

Removing them does not lose valid information.

Only a tiny fraction of data is affected. There are 1,746,661 total rows. Problematic rows:
* SO₂ negatives: 8,286
* CO negatives: 1,064
* Combined: < 0.5% of data

Removing them has zero statistical impact on AQI analysis.

Deleting invalid rows

In [43]:
df = df[(df["SO2 1st Max Value"] >= 0) & (df["CO Mean"] >= 0)]
print(df.shape)

(1737313, 28)


Below SO2 AQI column calculated based on given values in dedicated SO2 columns. 

In [44]:

def calculate_so2_aqi(C):
    """
    Calculate SO2 AQI using expanded breakpoint intervals (Option C).
    This avoids NA values from strict EPA bins.
    
    C : float 
        1-hour SO2 concentration in ppb
    """

    if pd.isna(C):
        return np.nan

    # ===== Expanded breakpoints ensuring continuous coverage =====
    if 0 <= C <= 35.999:
        Clow, Chigh = 0, 35
        Ilow, Ihigh = 0, 50

    elif 36 <= C <= 75.999:
        Clow, Chigh = 36, 75
        Ilow, Ihigh = 51, 100

    elif 76 <= C <= 185.999:
        Clow, Chigh = 76, 185
        Ilow, Ihigh = 101, 150

    elif 186 <= C <= 304.999:
        Clow, Chigh = 186, 304
        Ilow, Ihigh = 151, 200

    elif 305 <= C <= 604.999:
        Clow, Chigh = 305, 604
        Ilow, Ihigh = 201, 300

    else:
        # Out of range but we extend for safety
        return np.nan

    # ===== AQI Formula =====
    aqi = ((Ihigh - Ilow) / (Chigh - Clow)) * (C - Clow) + Ilow
    return round(aqi, 1)


In [45]:
# Function to add SO2 AQI column to DataFrame
def add_so2_aqi_column(df, col_name="SO2 1st Max Value"):
    """
    df : pandas dataframe  
    col_name : column containing SO2 1-hour max in ppb
    """
    df = df.copy()
    df["SO2 AQI"] = df[col_name].apply(calculate_so2_aqi)
    return df

In [46]:
df = add_so2_aqi_column(df, "SO2 1st Max Value")
df["SO2 AQI"].isna().sum()

np.int64(0)

And check column  and values

In [47]:
print(df[["SO2 1st Max Value", "SO2 AQI"]].head())

   SO2 1st Max Value  SO2 AQI
0                9.0     12.9
1                9.0     12.9
2                6.6      9.4
3                6.6      9.4
4                3.0      4.3


And CO AQI calculation

In [48]:
import numpy as np
import pandas as pd

def calculate_co_aqi(C):
    """
    Calculate CO AQI using expanded breakpoint intervals (Option C).
    This ensures continuous coverage with NO missing AQI values.

    C : float 
        8-hour CO concentration in ppm
    """

    if pd.isna(C):
        return np.nan

    # ===== Expanded breakpoints =====
    if 0.0 <= C <= 4.499:
        Clow, Chigh = 0.0, 4.4
        Ilow, Ihigh = 0, 50

    elif 4.5 <= C <= 9.499:
        Clow, Chigh = 4.5, 9.4
        Ilow, Ihigh = 51, 100

    elif 9.5 <= C <= 12.499:
        Clow, Chigh = 9.5, 12.4
        Ilow, Ihigh = 101, 150

    elif 12.5 <= C <= 15.499:
        Clow, Chigh = 12.5, 15.4
        Ilow, Ihigh = 151, 200

    elif 15.5 <= C <= 30.499:
        Clow, Chigh = 15.5, 30.4
        Ilow, Ihigh = 201, 300

    else:
        return np.nan  # extremely high or wrong units

    # ===== AQI Formula =====
    aqi = ((Ihigh - Ilow) / (Chigh - Clow)) * (C - Clow) + Ilow
    return round(aqi, 1)


In [49]:
def add_co_aqi_column(df, col="CO Mean"):
    df = df.copy()
    df["CO AQI"] = df[col].apply(calculate_co_aqi)
    return df

In [50]:
df = add_co_aqi_column(df, "CO Mean")
df["CO AQI"].isna().sum()

np.int64(0)

Verifying imputations

In [51]:
df.isnull().sum()           # Check for missing values in each column

State Code           0
County Code          0
Site Num             0
Address              0
State                0
County               0
City                 0
Date Local           0
NO2 Units            0
NO2 Mean             0
NO2 1st Max Value    0
NO2 1st Max Hour     0
NO2 AQI              0
O3 Units             0
O3 Mean              0
O3 1st Max Value     0
O3 1st Max Hour      0
O3 AQI               0
SO2 Units            0
SO2 Mean             0
SO2 1st Max Value    0
SO2 1st Max Hour     0
SO2 AQI              0
CO Units             0
CO Mean              0
CO 1st Max Value     0
CO 1st Max Hour      0
CO AQI               0
dtype: int64

And searching for negative valeus in other numeric columns:

In [52]:
def validate_no_negative_values(df, auto_fix=False, stop_on_error=False):
    """
    Validate that no negative values exist in any numeric column.
    
    auto_fix: If True → removes all rows containing negative values.
    stop_on_error: If True → raises an exception if negatives exist.
    """

    numeric_cols = df.select_dtypes(include=[np.number]).columns

    negatives_report = {}

    for col in numeric_cols:
        neg_mask = df[col] < 0
        neg_count = neg_mask.sum()

        if neg_count > 0:
            negatives_report[col] = {
                "count": int(neg_count),
                "sample": df.loc[neg_mask].head()
            }

    # If clean → report success
    if len(negatives_report) == 0:
        print("✅ No negative values found in ANY numeric column.")
        return df

    # If negatives exist → print detailed report
    print("❌ Negative values found!\n")

    for col, info in negatives_report.items():
        print(f"Column: {col}")
        print(f"  → Negative count: {info['count']}")
        print(f"  → Sample rows with negatives:")
        print(info["sample"])
        print("-" * 40)

    # Auto-fix option: remove rows
    if auto_fix:
        print("🧹 Auto-fix: Removing rows containing any negative values...")
        df = df[(df[numeric_cols] >= 0).all(axis=1)]
        print("✔ Negative rows removed.")
        return df

    # Stop execution option
    if stop_on_error:
        raise ValueError("Dataset contains negative values! See report above.")

    return df


In [53]:
validate_no_negative_values(df)
df = validate_no_negative_values(df, auto_fix=True)
df = validate_no_negative_values(df, stop_on_error=True)


❌ Negative values found!

Column: NO2 Mean
  → Negative count: 828
  → Sample rows with negatives:
         State Code  County Code  Site Num  \
1144580          23            3      1100   
1144581          23            3      1100   
1144582          23            3      1100   
1144583          23            3      1100   
1157402          36           55      1007   

                                         Address     State     County  \
1144580  8 NORTHERN ROAD, PRESQUE ISLE, ME 04769     Maine  Aroostook   
1144581  8 NORTHERN ROAD, PRESQUE ISLE, ME 04769     Maine  Aroostook   
1144582  8 NORTHERN ROAD, PRESQUE ISLE, ME 04769     Maine  Aroostook   
1144583  8 NORTHERN ROAD, PRESQUE ISLE, ME 04769     Maine  Aroostook   
1157402         2 YARMOUTH ROAD, RG&E Substation  New York     Monroe   

                 City  Date Local          NO2 Units  NO2 Mean  ...  \
1144580  Presque Isle  2011-04-30  Parts per billion -0.279167  ...   
1144581  Presque Isle  2011-04-30  Parts pe

828 negative valeus were found in NO2 mean columns. These rows were automatically deleted.

In [54]:
print(df.shape)                     # Print the shape of the DataFrame  

(1722197, 28)


Checking for TRUE duplicates.

In [55]:
def find_full_pollutant_duplicates(df):
    """
    Find duplicates using:
    - Date Local
    - Address
    - All pollutant mean / max / hour values
    - All AQI values
    
    Shows duplicates but DOES NOT remove them.
    """

    # Identify pollutant measurement columns dynamically
    pollutant_cols = [
        col for col in df.columns 
        if any(p in col for p in ["NO2", "O3", "SO2", "CO"])
    ]

    # Identify AQI columns
    aqi_cols = [col for col in df.columns if col.endswith("AQI")]

    # Build final set of columns for duplicate detection
    key_columns = ["Date Local", "Address"] + pollutant_cols + aqi_cols

    print("🔍 Checking duplicates using ALL pollutant columns:")
    print(key_columns, "\n")

    # Find duplicates (count both first and later occurrences)
    dup_mask = df.duplicated(subset=key_columns, keep=False)
    duplicates = df.loc[dup_mask].sort_values(by=key_columns)

    print(f"📌 Total FULL pollutant duplicates found: {len(duplicates)}\n")

    if len(duplicates) == 0:
        print("✅ No duplicates found.")
        return duplicates

    print("📄 Sample of duplicate rows (first 30):")
    display(duplicates.head(30))

    print("\n📊 Duplicate groups summary:")
    group_counts = (
        duplicates.groupby(["Date Local", "Address"])
        .size()
        .reset_index(name="Count")
        .sort_values("Count", ascending=False)
    )
    display(group_counts.head(20))

    return duplicates


In [56]:
duplicates = find_full_pollutant_duplicates(df)

🔍 Checking duplicates using ALL pollutant columns:
['Date Local', 'Address', 'NO2 Units', 'NO2 Mean', 'NO2 1st Max Value', 'NO2 1st Max Hour', 'NO2 AQI', 'O3 Units', 'O3 Mean', 'O3 1st Max Value', 'O3 1st Max Hour', 'O3 AQI', 'SO2 Units', 'SO2 Mean', 'SO2 1st Max Value', 'SO2 1st Max Hour', 'SO2 AQI', 'CO Units', 'CO Mean', 'CO 1st Max Value', 'CO 1st Max Hour', 'CO AQI', 'NO2 AQI', 'O3 AQI', 'SO2 AQI', 'CO AQI'] 

📌 Total FULL pollutant duplicates found: 94462

📄 Sample of duplicate rows (first 30):


State Code  County Code  Site Num  \
28240           6           83      1025   
28241           6           83      1025   
28242           6           83      1025   
28243           6           83      1025   
86568          51           59         5   
86569          51           59         5   
86566          51           59         5   
86567          51           59         5   
28246           6           83      1025   
28247           6           83      1025   
28244           6           83      1025   
28245           6           83      1025   
28254           6           83      1025   
28255           6           83      1025   
28252           6           83      1025   
28253           6           83      1025   
83460          48          141        37   
83461          48          141        37   
83462          48          141        37   
83463          48          141        37   
46168          20          107         2   
46169          20          107         2   
46166          20          107         2   
46167          20          107         2   
65736          36          103         9   
65737          36          103         9   
65734          36          103         9   
65735          36          103         9   
86636          51           59         5   
86637          51           59         5   

                                        Address       State         County  \
28240                  LFC #1-LAS FLORES CANYON  California  Santa Barbara   
28241                  LFC #1-LAS FLORES CANYON  California  Santa Barbara   
28242                  LFC #1-LAS FLORES CANYON  California  Santa Barbara   
28243                  LFC #1-LAS FLORES CANYON  California  Santa Barbara   
86568  CUB RUN LEE RD CHANT.(CUBRUN TREAT PLANT    Virginia        Fairfax   
86569  CUB RUN LEE RD CHANT.(CUBRUN TREAT PLANT    Virginia        Fairfax   
86566  CUB RUN LEE RD CHANT.(CUBRUN TREAT PLANT    Virginia        Fairfax   
86567  CUB RUN LEE RD CHANT.(CUBRUN TREAT PLANT    Virginia        Fairfax   
28246                  LFC #1-LAS FLORES CANYON  California  Santa Barbara   
28247                  LFC #1-LAS FLORES CANYON  California  Santa Barbara   
28244                  LFC #1-LAS FLORES CANYON  California  Santa Barbara   
28245                  LFC #1-LAS FLORES CANYON  California  Santa Barbara   
28254                  LFC #1-LAS FLORES CANYON  California  Santa Barbara   
28255                  LFC #1-LAS FLORES CANYON  California  Santa Barbara   
28252                  LFC #1-LAS FLORES CANYON  California  Santa Barbara   
28253                  LFC #1-LAS FLORES CANYON  California  Santa Barbara   
83460                                250 Rim Rd       Texas        El Paso   
83461                                250 Rim Rd       Texas        El Paso   
83462                                250 Rim Rd       Texas        El Paso   
83463                                250 Rim Rd       Texas        El Paso   
46168        COUNTY RD 1103 .7 MI SOUTH OF K-52      Kansas           Linn   
46169        COUNTY RD 1103 .7 MI SOUTH OF K-52      Kansas           Linn   
46166        COUNTY RD 1103 .7 MI SOUTH OF K-52      Kansas           Linn   
46167        COUNTY RD 1103 .7 MI SOUTH OF K-52      Kansas           Linn   
65736                        57 DIVISION STREET    New York        Suffolk   
65737                        57 DIVISION STREET    New York        Suffolk   
65734                        57 DIVISION STREET    New York        Suffolk   
65735                        57 DIVISION STREET    New York        Suffolk   
86636  CUB RUN LEE RD CHANT.(CUBRUN TREAT PLANT    Virginia        Fairfax   
86637  CUB RUN LEE RD CHANT.(CUBRUN TREAT PLANT    Virginia        Fairfax   

                City  Date Local          NO2 Units  NO2 Mean  ...  \
28240        Capitan  2000-01-03  Parts per billion  1.869565  ...   
28241        Capitan  2000-01-03  Parts per billion  1.869565  ...   
28242        


📊 Duplicate groups summary:


Date Local                       Address  Count
2471   2002-06-10                  HARRISON AVE    324
2465   2002-06-09                  HARRISON AVE    320
14416  2011-05-24  5888 MISSION BLVD., RUBIDOUX     80
16360  2012-10-11           2052 LAUWILIWILI ST     16
16608  2012-12-09           2052 LAUWILIWILI ST     16
16351  2012-10-08           2052 LAUWILIWILI ST     16
16010  2012-07-15           2052 LAUWILIWILI ST     16
20167  2015-01-25           2052 LAUWILIWILI ST     16
16173  2012-08-19           2052 LAUWILIWILI ST     16
16176  2012-08-20           2052 LAUWILIWILI ST     16
20330  2015-03-22           2052 LAUWILIWILI ST     16
14025  2011-02-09  5888 MISSION BLVD., RUBIDOUX     16
16697  2013-01-05           2052 LAUWILIWILI ST     16
16698  2013-01-06           2052 LAUWILIWILI ST     16
16815  2013-02-16           2052 LAUWILIWILI ST     16
16464  2012-10-30           2052 LAUWILIWILI ST     16
19103  2014-05-30           2052 LAUWILIWILI ST     16
16619  2012-12-11           2052 LAUWILIWILI ST     16
14315  2011-05-09  5888 MISSION BLVD., RUBIDOUX     16
12718  2010-03-14        PIKE AVE AT RIVER ROAD     16

Remove only TRUE Duplicates, perfectly matching.

This function:
* Removes only exact duplicates
*  Keeps the first occurrence
* Shows how many were removed
* Shows which Date/Address pairs had duplicates
* Returns a cleaned dataframe

In [57]:
def remove_full_pollutant_duplicates(df):
    """
    Remove ONLY true duplicates based on:
    - Date Local
    - Address
    - All pollutant measurement columns (Mean, Max, Hour, Units)
    - All AQI columns

    Keeps the FIRST occurrence.
    Returns cleaned dataframe + summary report.
    """

    df_clean = df.copy()

    # Identify pollutant columns dynamically
    pollutant_cols = [
        col for col in df.columns
        if any(p in col for p in ["NO2", "O3", "SO2", "CO"])
    ]

    # Identify AQI columns
    aqi_cols = [col for col in df.columns if col.endswith("AQI")]

    key_columns = ["Date Local", "Address"] + pollutant_cols + aqi_cols

    print("🧹 Removing true duplicates based on columns:")
    print(key_columns, "\n")

    before = len(df_clean)

    # Remove duplicates (keep first occurrence)
    df_clean = df_clean.drop_duplicates(subset=key_columns, keep="first")

    after = len(df_clean)
    removed = before - after

    print(f"📉 Total rows BEFORE: {before}")
    print(f"📈 Total rows AFTER:  {after}")
    print(f"🗑️ Removed duplicates: {removed}\n")

    # Show top duplicate groups (optional)
    if removed > 0:
        print("📊 Duplicate groups (Date + Address) impacted:")
        dup_groups = (
            df[df.duplicated(subset=key_columns, keep=False)]
            .groupby(["Date Local", "Address"])
            .size()
            .reset_index(name="Count")
            .sort_values("Count", ascending=False)
        )
        display(dup_groups.head(20))
    else:
        print("✅ No duplicates were removed. Dataset was already clean.")

    return df_clean


In [58]:
df = remove_full_pollutant_duplicates(df)


🧹 Removing true duplicates based on columns:
['Date Local', 'Address', 'NO2 Units', 'NO2 Mean', 'NO2 1st Max Value', 'NO2 1st Max Hour', 'NO2 AQI', 'O3 Units', 'O3 Mean', 'O3 1st Max Value', 'O3 1st Max Hour', 'O3 AQI', 'SO2 Units', 'SO2 Mean', 'SO2 1st Max Value', 'SO2 1st Max Hour', 'SO2 AQI', 'CO Units', 'CO Mean', 'CO 1st Max Value', 'CO 1st Max Hour', 'CO AQI', 'NO2 AQI', 'O3 AQI', 'SO2 AQI', 'CO AQI'] 

📉 Total rows BEFORE: 1722197
📈 Total rows AFTER:  1674675
🗑️ Removed duplicates: 47522

📊 Duplicate groups (Date + Address) impacted:


Date Local                       Address  Count
2471   2002-06-10                  HARRISON AVE    324
2465   2002-06-09                  HARRISON AVE    320
14416  2011-05-24  5888 MISSION BLVD., RUBIDOUX     80
16360  2012-10-11           2052 LAUWILIWILI ST     16
16608  2012-12-09           2052 LAUWILIWILI ST     16
16351  2012-10-08           2052 LAUWILIWILI ST     16
16010  2012-07-15           2052 LAUWILIWILI ST     16
20167  2015-01-25           2052 LAUWILIWILI ST     16
16173  2012-08-19           2052 LAUWILIWILI ST     16
16176  2012-08-20           2052 LAUWILIWILI ST     16
20330  2015-03-22           2052 LAUWILIWILI ST     16
14025  2011-02-09  5888 MISSION BLVD., RUBIDOUX     16
16697  2013-01-05           2052 LAUWILIWILI ST     16
16698  2013-01-06           2052 LAUWILIWILI ST     16
16815  2013-02-16           2052 LAUWILIWILI ST     16
16464  2012-10-30           2052 LAUWILIWILI ST     16
19103  2014-05-30           2052 LAUWILIWILI ST     16
16619  2012-12-11           2052 LAUWILIWILI ST     16
14315  2011-05-09  5888 MISSION BLVD., RUBIDOUX     16
12718  2010-03-14        PIKE AVE AT RIVER ROAD     16

Conflicting duplicates identification:
* Detect measurement inconsistencies
* Identify stations that report multiple measurements at same time
* Flag data quality problems
* Decide whether to:
 * average conflicts
 * drop the worst sensors
 * keep the max value (common for AQI rules)

Resolve Conflicting Duplicates. Keep the Maximum Values (EPA-style AQI logic)

EPA rules for AQI calculations already require maxima for 1-hour & 8-hour values.
So to remain consistent, we choose the maximum values within each conflict group.

* Best for AQI
* Prevents underestimation
* Officially aligned with U.S. EPA methodology

It ensures:
* proper pollutant selection
* AQI is never underestimated
* dataset integrity for environmental analysis

In [59]:
def detect_and_resolve_conflicts(df):
    df_copy = df.copy()

    key_cols = ["Date Local", "Address"]

    pollutant_cols = [c for c in df.columns if any(p in c for p in ["NO2", "O3", "SO2", "CO"])]
    aqi_cols = [c for c in df.columns if c.endswith("AQI")]
    group_cols = list(set(pollutant_cols + aqi_cols))

    print("\n🔍 Checking columns:")
    print("Pollutant cols:", pollutant_cols)
    print("AQI cols:", aqi_cols)
    print("Keys:", key_cols)

    # STEP 1 – identify conflicting groups
    grouped = (
        df_copy.groupby(key_cols)[group_cols]
        .nunique(dropna=False)
        .reset_index()
    )

    grouped["conflict"] = grouped[group_cols].max(axis=1) > 1
    conflict_keys = grouped[grouped["conflict"]][key_cols]

    print(f"\n⚠ Total conflicting groups found: {len(conflict_keys)}")
    if conflict_keys.empty:
        print("✓ No conflicts detected.")
        return df_copy, pd.DataFrame()

    # STEP 2 – extract all conflicting rows
    conflicts = df_copy.merge(conflict_keys, on=key_cols, how="inner")
    print(f"⚠ Total conflicting rows: {len(conflicts)}")

    # STEP 3 – resolve conflicts using MAX
    resolved_conflicts = (
        conflicts.groupby(key_cols)[group_cols]
        .max()
        .reset_index()
    )

    # STEP 4 – get all non-conflicting rows
    non_conflicts = df_copy.merge(conflict_keys, on=key_cols, how="left", indicator=True)
    non_conflicts = non_conflicts[non_conflicts["_merge"] == "left_only"].drop(columns="_merge")

    # STEP 5 – align column order
    resolved_conflicts = resolved_conflicts.reindex(columns=df_copy.columns, fill_value=np.nan)

    # STEP 6 – reset index to avoid "duplicate labels" error
    resolved_conflicts = resolved_conflicts.reset_index(drop=True)
    non_conflicts = non_conflicts.reset_index(drop=True)

    # STEP 7 – final dataset
    df_cleaned = pd.concat([non_conflicts, resolved_conflicts], ignore_index=True)

    print("\n✓ Conflicts resolved using MAX pollutant values.")
    print(f"✓ Clean rows: {len(non_conflicts)}, Resolved rows: {len(resolved_conflicts)}")
    print(f"✓ Final dataset size: {len(df_cleaned)}")

    return df_cleaned, conflicts


In [60]:
df_cleaned, conflicts = detect_and_resolve_conflicts(df)



🔍 Checking columns:
Pollutant cols: ['NO2 Units', 'NO2 Mean', 'NO2 1st Max Value', 'NO2 1st Max Hour', 'NO2 AQI', 'O3 Units', 'O3 Mean', 'O3 1st Max Value', 'O3 1st Max Hour', 'O3 AQI', 'SO2 Units', 'SO2 Mean', 'SO2 1st Max Value', 'SO2 1st Max Hour', 'SO2 AQI', 'CO Units', 'CO Mean', 'CO 1st Max Value', 'CO 1st Max Hour', 'CO AQI']
AQI cols: ['NO2 AQI', 'O3 AQI', 'SO2 AQI', 'CO AQI']
Keys: ['Date Local', 'Address']

⚠ Total conflicting groups found: 407434
⚠ Total conflicting rows: 1674552

✓ Conflicts resolved using MAX pollutant values.
✓ Clean rows: 123, Resolved rows: 407434
✓ Final dataset size: 407557


In [61]:
df_cleaned.head()

State Code  County Code  Site Num                           Address  \
0        36.0        103.0       9.0                57 DIVISION STREET   
1         6.0         71.0     306.0  14306 PARK AVE., VICTORVILLE, CA   
2        22.0         33.0       9.0              1061-A Leesville Ave   
3         6.0         37.0    1103.0       1630 N MAIN ST, LOS ANGELES   
4         6.0         37.0    5005.0       7201 W. WESTCHESTER PARKWAY   

        State            County         City  Date Local          NO2 Units  \
0    New York           Suffolk   Holtsville  2000-10-09  Parts per billion   
1  California    San Bernardino  Victorville  2001-02-17  Parts per billion   
2   Louisiana  East Baton Rouge  Baton Rouge  2002-05-03  Parts per billion   
3  California       Los Angeles  Los Angeles  2004-04-24  Parts per billion   
4  California       Los Angeles  Los Angeles  2004-05-24  Parts per billion   

    NO2 Mean  ...          SO2 Units  SO2 Mean  SO2 1st Max Value  \
0   7.750000  ...  Parts per billion       1.0                2.0   
1  19.000000  ...  Parts per billion       1.0                1.0   
2  10.583333  ...  Parts per billion       0.0                0.0   
3  50.000000  ...  Parts per billion       2.0                2.0   
4   3.090909  ...  Parts per billion       1.0                1.0   

  SO2 1st Max Hour  SO2 AQI           CO Units  CO Mean  CO 1st Max Value  \
0               10      2.9  Parts per million      0.0               0.0   
1               22      1.4  Parts per million      0.5               0.5   
2                0      0.0  Parts per million      0.0               0.0   
3                0      2.9  Parts per million      1.0               1.0   
4                6      1.4  Parts per million      0.2               0.2   

  CO 1st Max Hour  CO AQI  
0               0     0.0  
1              22     5.7  
2               0     0.0  
3               0    11.4  
4               0     2.3  

[5 rows x 28 columns]

In [62]:
conflicts.head()

State Code  County Code  Site Num                                  Address  \
0           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN   
1           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN   
2           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN   
3           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN   
4           4           13      3002  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN   

     State    County     City  Date Local          NO2 Units   NO2 Mean  ...  \
0  Arizona  Maricopa  Phoenix  2000-01-01  Parts per billion  19.041667  ...   
1  Arizona  Maricopa  Phoenix  2000-01-01  Parts per billion  19.041667  ...   
2  Arizona  Maricopa  Phoenix  2000-01-01  Parts per billion  19.041667  ...   
3  Arizona  Maricopa  Phoenix  2000-01-01  Parts per billion  19.041667  ...   
4  Arizona  Maricopa  Phoenix  2000-01-02  Parts per billion  22.958333  ...   

           SO2 Units  SO2 Mean  SO2 1st Max Value SO2 1st Max Hour  SO2 AQI  \
0  Parts per billion  3.000000                9.0               21     12.9   
1  Parts per billion  3.000000                9.0               21     12.9   
2  Parts per billion  2.975000                6.6               23      9.4   
3  Parts per billion  2.975000                6.6               23      9.4   
4  Parts per billion  1.958333                3.0               22      4.3   

            CO Units   CO Mean  CO 1st Max Value CO 1st Max Hour  CO AQI  
0  Parts per million  1.145833               4.2              21    13.0  
1  Parts per million  0.878947               2.2              23    10.0  
2  Parts per million  1.145833               4.2              21    13.0  
3  Parts per million  0.878947               2.2              23    10.0  
4  Parts per million  0.850000               1.6              23     9.7  

[5 rows x 28 columns]

In [63]:
df.shape

(1674675, 28)

In [ ]:
from geopy.geocoders import Nominatim
import pandas as pd
import time

geolocator = Nominatim(user_agent="my_app")

def geocode_address(address):
    try:
        location = geolocator.geocode(address)
        if location:
            return location.latitude, location.longitude
    except:
        return None, None

df["lat"], df["lon"] = zip(*df["Address"].apply(geocode_address))
time.sleep(1)  # avoid rate limit


In [ ]:
import folium

m = folium.Map(location=[df.lat.mean(), df.lon.mean()], zoom_start=10)

for _, row in df.iterrows():
    folium.Marker(
        location=[row.lat, row.lon],
        popup=row.Address
    ).add_to(m)

m
